### PR TITLE
Improving BuildDistribution

### PR DESCRIPTION
**Pull request by sofianehaddad**
The idea of the following PR is to improve the usage of `FunctionalChaosAlgorithm::BuildDistribution`
For that purpose, we first added the `AIC` & `AICc` eval in `FittingTest`. The usage of such methods is not restrictive to the chaos of course.
Then we moved the `static` method from `FunctionalChaosAlgorithm` to `MetaModelAlgorithm`.

Finally we update the method. The proposal is as follows:
 - From a list of factories, we select the potential `candidates` (i.e underlying distribution is estimated and its associated `p-value` is higher then the level)
 - We filter the previous candidates & select the one with the best criterion (`BIC`, `AIC`, `AICc`, `p-value`)
 - If there is no candidates (`p-value` low) we use a non parametric model

### CODE REVIEW

File: lib/src/Uncertainty/Algorithm/MetaModel/MetaModelAlgorithm.cxx [hunk 1]
```diff
@@ -67,47 +68,124 @@ String MetaModelAlgorithm::__repr__() const
...
-  KernelSmoothing ks;
+  // The strategy for the marginals is to find the best continuous 1-d parametric model else fallback to a non-parametric one
+  DistributionFactory nonParametricModel;
+  const String nonParametric = ResourceMap::GetAsString("MetaModelAlgorithm-NonParametricModel");
+  if (nonParametric == "Histogram")
+    nonParametricModel = DistributionFactory(HistogramFactory());
+  else
+    nonParametricModel = DistributionFactory(KernelSmoothing());
+  //Collection< DistributionFactory > factories = {NormalFactory(), UniformFactory(), GammaFactory(), ExponentialFactory()}
   Collection< DistributionFactory > factories(DistributionFactory::GetContinuousUniVariateFactories());
```
Comment (jschueller) [inline lib/src/Uncertainty/Algorithm/MetaModel/MetaModelAlgorithm.cxx:123]: GetContinuousUniVariateFactories also return Histogram, so you might want to exclude it

**Comment by jschueller** [comment]
I started to work on this in #1588, but this is more advanced, i'll remove my commits

**Comment by jschueller** [comment]
please mention #1415

**Comment by sofianehaddad** [comment]
@jschueller  I used some elements from your [fix](https://github.com/openturns/openturns/commit/b652f782b08557230cd168fd47b53851361b1d63) provided by 

We might have `DistributionFactory::GetXXXFactories` specific for  metamodels but the question is how to update/feed this method outside ?

Ok for the ticket

**Comment by jschueller** [comment]
I think we want almost all distritbutions except histogram or the few costly ones, so I just filtered the result of GetContinuousUniVariateFactories.

File: lib/src/Uncertainty/Algorithm/MetaModel/FunctionalChaos/FunctionalChaosAlgorithm.cxx [hunk 2]
```diff
@@ -153,55 +147,6 @@ FunctionalChaosAlgorithm::FunctionalChaosAlgorithm(const Sample & inputSample,
   if (inputSample.getSize() != outputSample.getSize()) throw InvalidArgumentException(HERE) << "Error: the input sample and the output sample must have the same size.";
 }
 
-Distribution FunctionalChaosAlgorithm::BuildDistribution(const Sample & inputSample)
```
Comment (jschueller) [inline lib/src/Uncertainty/Algorithm/MetaModel/FunctionalChaos/FunctionalChaosAlgorithm.cxx:156]: some testing:
```
#! /usr/bin/env python

from __future__ import print_function
import openturns as ot
from math import pi

ot.TESTPREAMBLE()
ot.Log.Show(0)
ot.PlatformInfo.SetNumericalPrecision(3)


dim = 2
R=ot.CorrelationMatrix(2)
R[0,1] = 0.5
src = [ot.ComposedDistribution([ot.Uniform(-pi, pi)] * dim),
       ot.ComposedDistribution([ot.Normal(4.0,2.0)] * dim),
       ot.ComposedDistribution([ot.Gamma()] * dim),
       ot.ComposedDistribution([ot.Gamma(1.5, 2.5, -0.5)] * dim),
       ot.ComposedDistribution([ot.Arcsine(5.2, 11.6)] * dim),
       ot.Normal([3.0]*dim, [2.0]*dim, R)]

for sd in src:
    sample = sd.getSample(10000)
    d = ot.MetaModelAlgorithm.BuildDistribution(sample)
    print(d)
```

Comment (sofianehaddad) [reply lib/src/Uncertainty/Algorithm/MetaModel/FunctionalChaos/FunctionalChaosAlgorithm.cxx:156]: @jschueller thanks. I included the test

File: lib/src/Uncertainty/Algorithm/MetaModel/MetaModelAlgorithm.cxx [hunk 3]
```diff
@@ -60,6 +68,141 @@ String MetaModelAlgorithm::__repr__() const
...
+
+  const Description inputDescription(inputSample.getDescription());
+  for (UnsignedInteger i = 0; i < inputDimension; ++i)
+  {
+    // Here we remove the duplicate entries in the marginal sample as we are suppose to have a continuous distribution. 
+    // The duplicates are mostly due to truncation in the file export.
+    const Sample marginalSample(inputSample.getMarginal(i).sortUnique());
+    // First we estimate distribution using its factory
+    // Once it is possible, we keep it only if Kolmogorov score exceed level
+    std::vector<TestedDistribution> possibleDistributions;
```
Comment (regislebrun) [inline lib/src/Uncertainty/Algorithm/MetaModel/MetaModelAlgorithm.cxx:147]: Why do you use a std::vector instead of a Collection here?

**Review by regislebrun — APPROVED** [review]
Well done!

File: lib/src/Uncertainty/Algorithm/MetaModel/MetaModelAlgorithm.cxx (hunk 3 above)
Comment (sofianehaddad) [reply lib/src/Uncertainty/Algorithm/MetaModel/MetaModelAlgorithm.cxx:147]: I do not remember exactly the issue I had. So I moved to `std::vector`

File: python/src/MetaModelAlgorithm_doc.i.in [hunk 4]
```diff
@@ -21,6 +21,41 @@ A MetaModelAlgorithm object can be used only through its derived classes:
...
+  - We estimate the parameters of each candidate when feasible. If so, the model
+  is said to be `valid` if its Kolmogorov score exceeds a threshold fixed in the
+  `MetaModelAlgorithm-PValueThreshold` ResourceMap key. Default value is 5%
+  - We sort the `valid` models and return the one with the optimal criterion. Criterion might
+  be `BIC`, `AIC` or `AICC` (specification through the `MetaModelAlgorithm-ModelSelectionCriterion`
+  ResourceMap key, with default value fixed to `BIC`)
+  - If there is no `valid` candidate, we estimate a non-parametric model (:class:`~openturns.KernelSmoothing`
+  or :class:`~openturns.Histogram`). The `MetaModelAlgorithm-NonParametricModel` ResourceMap key allows selecting
+  the preferred one. Default value is `Histogram`
+
```
Comment (jschueller) [inline python/src/MetaModelAlgorithm_doc.i.in:44]: you'got to indent the text below each bullet

File: python/src/MetaModelAlgorithm_doc.i.in [hunk 5]
```diff
@@ -21,6 +21,41 @@ A MetaModelAlgorithm object can be used only through its derived classes:
...
 // ---------------------------------------------------------------------
 
+
+%feature("docstring") OT::MetaModelAlgorithm::BuildDistribution
+"Recover the distribution, with metamodel performance in mind.
+
+For each marginal, find the best 1-d continuous parametric model
+else fallback to the use of a nonparametric one.
+
+The selection is done as follow :
```
Comment (jschueller) [inline python/src/MetaModelAlgorithm_doc.i.in:31]: follow:

**Review by jschueller — APPROVED** [review]

**Comment by jschueller** [comment]
a rebase is needed

**Comment by sofianehaddad** [comment]
Done

**Comment by jschueller** [comment]
thanks

File: python/src/MetaModelAlgorithm_doc.i.in [hunk 6]
```diff
@@ -21,6 +21,41 @@ A MetaModelAlgorithm object can be used only through its derived classes:
...
+
+For each marginal, find the best 1-d continuous parametric model
+else fallback to the use of a nonparametric one.
+
+The selection is done as follow:
+
+  - We start with a list of all parametric models (all factories)
+  - For each model, we estimate its parameters if feasible.
+  - We check then if model is `valid`, ie if its Kolmogorov score exceeds a threshold 
+  fixed in the `MetaModelAlgorithm-PValueThreshold` ResourceMap key. Default value is 5%
```
Comment (jschueller) [inline python/src/MetaModelAlgorithm_doc.i.in:36]: missing  indent here

**Pull request merged**